homeassistant/components/smappee/config_flow.py

```diff
@@ -21,3 +28,123 @@ class SmappeeFlowHandler(
     def logger(self) -> logging.Logger:
         """Return logger."""
         return logging.getLogger(__name__)
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        return await self._handle_config_flow(user_input)
+
+    async def async_step_zeroconf(self, user_input):
+        """Handle zeroconf discovery."""
+        if user_input is None:
+            return self.async_abort(reason="connection_error")
+
+        if not user_input[CONF_HOSTNAME].startswith("Smappee1"):
+            # We currently only support Energy and Solar models (legacy)
+            return self.async_abort(reason="invalid_mdns")
+
+        self.context.update(
+            {
+                CONF_IP_ADDRESS: user_input["host"],
+                CONF_SERIALNUMBER: user_input[CONF_HOSTNAME]
+                .replace(".local.", "")
+                .replace("Smappee", ""),
+            }
+        )
+
+        # Prepare configuration flow
+        return await self._handle_config_flow(user_input, True)
+
+    async def async_step_zeroconf_confirm(self, user_input=None):
+        """Handle a flow initiated by zeroconf."""
+        return await self._handle_config_flow(user_input)
+
+    async def _handle_config_flow(self, user_input=None, prepare=False):
+        """Config flow handler for Smappee."""
+        source = self.context.get("source")
+
+        # Show LOCAL/CLOUD option or prepare zeroconf discovery flow
+        if user_input is None and not prepare:
+            if source == SOURCE_ZEROCONF:
+                return self._show_zeroconf_confirm_dialog()
+
+            # Show environment form with CLOUD or LOCAL option
+            return self._show_setup_form(step="environment")
+
+        # Environment chosen, request additional host information for LOCAL or OAuth2 flow for CLOUD
+        if user_input is not None and "environment" in user_input:
```

Comment:
All these checks makes the flow really hard to follow. Please create one step method for each form and put the logic for the step inside the step method. Helper methods are usually only needed for the api interacting logic that may be common for different flow paths.

Reply:
So everything from the _handle_config_flow method should be moved to the async_step_user and async_step_zeroconf methods?

Reply:
Because the integration can now be added multiple ways (yaml configuration, zeroconf discovery, manually entering the host) it seemed like a good way to follow the WLED config flow https://github.com/home-assistant/core/blob/dev/homeassistant/components/wled/config_flow.py#L64:L108? Would it be possible to explain the requested changes you want a bit more or maybe make a proposal?

Reply:
I'll have time give more feedback in the second half of this week.

Reply:
Thanks! I already made an attempt but quite unclear if it's the way you described.